src/material.cpp

```diff
@@ -469,13 +485,163 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
   H5Sclose(metadatapace);
   H5Tclose(attrtype);
 
+  delete[] mat_data;
+
+}
+
+
+void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+                                float row, int chunksize) {
+  hid_t material_grp_id;  // Holder of HDF5 Id of the "/material" group
+  hid_t data_id;  // Holder of HDF5 Id of the data group to write the data
+                  // (located in "/material/datapath")
+
+  // Turn off annoying HDF5 errors
+  H5Eset_auto2(H5E_DEFAULT, NULL, NULL);
+
+  // Set file access properties so it closes cleanly
+  hid_t fapl;
+  fapl = H5Pcreate(H5P_FILE_ACCESS);
+  H5Pset_fclose_degree(fapl, H5F_CLOSE_STRONG);
+  // Create new/open datafile.
+
+  // This complicated algorythm is required to allow backward compatibility with
+  // previous version of write_hdf5 (where data were written in a hdf5 DATASET)
+  // FILE EXIST ?
+  //    NO -> create file with a "/material" group, write the data in it
+  //    YES:
+  //      - DATAPTH exist:
+  //        - YES && != /material -> Detect NUCPATH + old write_hdf5
+  //        - NO -> CONTINUE
+  //          - "/material" EXIST:
+  //            - NO -> create "/material" group and write everyting in it
+  //            - YES:
+  //              - "/material" is a DATASET -> detect NUCPATH + old write_hdf5
+  //              - "/material" is a GROUP -> add data in it
+  //              - "/material" isn't a GROUP nor a DATASET -> fail and complain
+
+  hid_t db;
+  if (!pyne::file_exists(filename)) {
+    // Create the file
+    db = H5Fcreate(filename.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, fapl);
+
+  } else {
+    bool ish5 = H5Fis_hdf5(filename.c_str());
+    if (!ish5) throw h5wrap::FileNotHDF5(filename);
+    db = H5Fopen(filename.c_str(), H5F_ACC_RDWR, fapl);
+
+    // if datapath != /material and exist -> old format
+    if (h5wrap::path_exists(db, datapath) && datapath != "/material") {
```

Comment:
I think this might need an additional comment.

If the user gives `datapath = "my_mat"` this `path_exists()` will definitely fail, right?  But if the user gives `datapath = "/my_mat"` then it may succeed as an instance of the old format.  Furthermore, the latter example is never valid in our new structure.  I think this leading slash is a very subtle difference that is worth explaining.

Reply:
I forgot to add the `/` into the `datapath` is not already present...


 it should resolve this potential issue

Reply:
I think it’s probably correct, but it took me a long time of thinking carefully about the difference between the cases I mentioned to convince myself of the behavior